### PR TITLE
feature/server-client-protocol-version-2

### LINKED ISSUE #905: Q: How is reading from a Service TCP Socket handled in eCAL?
This question is about how the data-read from the TCP Socket is handled in eCAL, when a service has been called. So I am talking about the **Service Server** side.

First of all: here is a screenshot of a call stack:
image

This question explicitly goes to @rex-schilasky. Please confirm or correct me, if I am wrong.

# My understanding

From what I understand, the logic is as follows:

1. A random number of bytes is read from the Socket
    https://github.com/eclipse-ecal/ecal/blob/d78e95b1c27dd99872474d6ba9b4c382c9013c78/ecal/core/src/service/asio_server.h#L54-L60

2. If there is a callback, the data is added to a member variable. I think there always is a request callback.
    https://github.com/eclipse-ecal/ecal/blob/d78e95b1c27dd99872474d6ba9b4c382c9013c78/ecal/core/src/service/asio_server.h#L78-L82
    https://github.com/eclipse-ecal/ecal/blob/d78e95b1c27dd99872474d6ba9b4c382c9013c78/ecal/core/src/service/asio_server.h#L170

3. It is checked if there is any more data available on the socket. If there is, again a random number of bytes is read. The data is added to the `request_` variable forever or until all currently available bytes are read
    https://github.com/eclipse-ecal/ecal/blob/d78e95b1c27dd99872474d6ba9b4c382c9013c78/ecal/core/src/service/asio_server.h#L84-L91

4. If there are no more bytes available, the callback is executed with whatever has been accumulated since the last call. Then the accumulated `request_` is cleared.
   https://github.com/eclipse-ecal/ecal/blob/d78e95b1c27dd99872474d6ba9b4c382c9013c78/ecal/core/src/service/asio_server.h#L98-L99

# The issue

If it is true what I wrote here, the Service API is inherently unstable and unsafe
- There seems to be no protocol involved telling how many bytes a message has
- Multiple messages can stack up in the TCP buffer and are interpreted as one contiguous message. There is no way of knowing.
- The pure absense of data is immediatelly interpreted as "end-of-message". The code always requests a random number of bytes from the socket, so when the TCP Stream is empty e.g. because an ethernet frame has been delayed or had to be re-transmitted, the message is cut into pieces and the next chunks are interpreted as their own messages or are used to prefix the next message, that may then be available.

### LINKED ISSUE #907: Unreliable RPC calls with large messages
**Describe the bug**
We are sending some long strings in protobuf RPC calls around and it seems like the calls get lost more often with increasing message size. This just seems to be happening when communicating over an ethernet network. I couldn't reproduce it on localhost communication.

I think it is related to https://github.com/eclipse-ecal/ecal/issues/905

This might also be related to https://github.com/eclipse-ecal/ecal/issues/581 but probably doesn't happen that often in a docker setup.

**To Reproduce**
1. Patch the example ping rpc application:

I have set the size of the string data to something similar we are trying to use in our application. It's really just a small base64 encoded image. The size is likely to be a bit larger in the future.

```
diff --git a/samples/cpp/services/ping_client/src/ping_client.cpp b/samples/cpp/services/ping_client/src/ping_client.cpp
index 75ab654e..06d111ed 100644
--- a/samples/cpp/services/ping_client/src/ping_client.cpp
+++ b/samples/cpp/services/ping_client/src/ping_client.cpp
@@ -51,10 +51,11 @@ int main(int argc, char **argv)
       //////////////////////////////////////
       PingRequest               ping_request;
       eCAL::ServiceResponseVecT service_response_vec;
-      ping_request.set_message("PING");
+      ping_request.set_message(std::string(25900, 'a'));
+      std::cout << "ping_request.message().size(): " << ping_request.message().size() << std::endl;
       if (ping_client.Call("Ping", ping_request, -1, &service_response_vec))
       {
-        std::cout << std::endl << "PingService::Ping method called with message : " << ping_request.message() << std::endl;
+        std::cout << std::endl << "PingService::Ping method called with message : " << ping_request.message().size() << std::endl;
 
         for (auto service_response : service_response_vec)
         {
@@ -84,7 +85,7 @@ int main(int argc, char **argv)
     }
 
     // sleep a second
-    std::this_thread::sleep_for(std::chrono::milliseconds(1000));
+    std::this_thread::sleep_for(std::chrono::milliseconds(100));
   }
 
   // finalize eCAL API
diff --git a/samples/cpp/services/ping_server/src/ping_server.cpp b/samples/cpp/services/ping_server/src/ping_server.cpp
index 1973d57d..c87349f9 100644
--- a/samples/cpp/services/ping_server/src/ping_server.cpp
+++ b/samples/cpp/services/ping_server/src/ping_server.cpp
@@ -35,9 +35,9 @@ public:
   void Ping(::google::protobuf::RpcController* /* controller_ */, const ::PingRequest* request_, ::PingResponse* response_, ::google::protobuf::Closure* /* done_ */)
   {
     // print request and
-    std::cout << "Received request PingService / Ping     : " << request_->message() << std::endl << std::endl;
+    std::cout << "Received request PingService / Ping     : " << request_->message().size() << std::endl << std::endl;
     // create response
-    response_->set_answer("PONG");
+    response_->set_answer(std::to_string(request_->message().size()));
   }
 };
```

3. Run server and client on two different machines connected via ethernet. Set up 239.0.0.0 routes, hostsnames, etc...
4. A lot of RPC calls seem to fail. At least on my setup.  

```
~/.../bin (master)$ ./ecal_sample_ping_client
Waiting for the service ..
ping_request.message().size(): 25900
PingService::Ping method call failed ..

ping_request.message().size(): 25900

PingService::Ping method called with message : 25900
Received response PingService / Ping : 25900 from host ubuntu-X570S-UD
ping_request.message().size(): 25900

PingService::Ping method called with message : 25900
Received response PingService / Ping : 25900 from host ubuntu-X570S-UD
ping_request.message().size(): 25900
PingService::Ping method call failed ..

ping_request.message().size(): 25900
PingService::Ping method call failed ..

ping_request.message().size(): 25900
PingService::Ping method call failed ..

ping_request.message().size(): 25900

PingService::Ping method called with message : 25900
Received response PingService / Ping : 25900 from host ubuntu-X570S-UD
ping_request.message().size(): 25900
PingService::Ping method call failed ..

ping_request.message().size(): 25900

PingService::Ping method called with message : 25900
Received response PingService / Ping : 25900 from host ubuntu-X570S-UD
ping_request.message().size(): 25900

PingService::Ping method called with message : 25900
Received response PingService / Ping : 25900 from host ubuntu-X570S-UD
ping_request.message().size(): 25900
PingService::Ping method call failed ..

```

**Expected behavior**
Service calls shouldn't randomly fail.

**Desktop (please complete the following information):**
 - OS: Ubuntu
 - Version: 22.04

### PR DESCRIPTION
**Pull request by rex-schilasky**
**Pull request type**

Please check the type of change your PR introduces:
- [X] Bugfix

**What is the current behavior?**
See issue #907 and pull request #918.

Fixes #905
Fixes #907

**What is the new behavior?**
Different client/server protocol versions introduced based on the work of @floroeske. 

1. Client and server exchange their protocol version via registration layer.
2. Server will not start before any client registration is received.
3. Server starts with clients protocol version (currently 0 or 1).
4. A server can handle multiple clients with different protocol version.
5. A client can be connected to multiple server with different protocol versions.
 
**Does this introduce a breaking change?**

- [X] Yes

Public API changed /include/ecal/ecal_callback.h.

### CODE REVIEW

File: ecal/core/src/service/asio_server.h [hunk 1]
```diff
@@ -125,6 +140,84 @@ class CAsioSession
...
+          memcpy(&tcp_header, header_.data(), sizeof(eCAL::STcpHeader));
+          header_request_size_ = static_cast<size_t>(ntohl(tcp_header.psize_n));
+        }
+
+        // collect request_
+        if (header_request_size_ != 0 && request_.size() < header_request_size_)
+        {
+          request_ += data.substr(bytes_used, bytes_transferred - bytes_used);
+        }
+
```
Comment (KanonWY) [inline ecal/core/src/service/asio_server.h:177]: Consider a relatively rare case, suppose pack_1 and pack_2 are sent in rapid succession on the tcpclient side. If pack_1 is internally split for some reason, and then resend part of the internally split packets (tcp automatic retransmission), then the new packet actually contains the second half of pack_1 and the first half of pack_2, the current code implementation does not do anything to the first half of pack_2.
Of course, due to the nodelay socket option enabled on the sender side, this makes it difficult to reproduce the situation, but there is still a low probability of this problem.

File: ecal/core/src/service/ecal_tcpclient.cpp [hunk 2]
```diff
@@ -241,7 +316,7 @@ namespace eCAL
       // read stream header (sync)
       else
       {
-        size_t bytes_read = m_socket->read_some(asio::buffer(&tcp_header, sizeof(tcp_header)));
+        size_t const bytes_read = m_socket->read_some(asio::buffer(&tcp_header, sizeof(tcp_header)));
```
Comment (floroeske) [inline ecal/core/src/service/ecal_tcpclient.cpp:319]: Maybe same thing here. Validate header and check if psize_n makes sense.

File: ecal/core/src/service/ecal_tcpclient.cpp [hunk 3]
```diff
@@ -196,6 +224,53 @@ namespace eCAL
...
+    {
+      std::cerr << "CTcpClient::SendRequest: Failed to send request: " << e.what() << "\n";
+      m_connected = false;
+      return false;
+    }
+
+    return true;
+  }
+
+  std::vector<char> CTcpClient::PackRequest(const std::string& request)
```
Comment (floroeske) [inline ecal/core/src/service/ecal_tcpclient.cpp:260]: Wondering if PackRequest and PackResponse should live in core/src/service/ecal_tcpheader.h. Maybe one PackHeader() function would be enough.

**Review by floroeske — COMMENTED** [review]
Cool, I'll try it out in the next few days.

File: ecal/core/src/service/asio_server.h (hunk 1 above)
Comment (rex-schilasky) [reply ecal/core/src/service/asio_server.h:177]: First of all, thank you for your review - GREAT. I will check all in detail later.

**Comment by floroeske** [comment]
I've been running this branch for the last week on some of our devices and I haven't found any issues so far. I haven't tested the protocol backwards compatibility.

**Comment by FlorianReimold** [comment]
I am finally handing this over to anybody who would like to review the changes 😉

**Review by hannemn — APPROVED** [review]

**Pull request merged**